nixos/modules/services/logging/beats/lib.nix

```diff
@@ -0,0 +1,145 @@
+{ lib, pkgs, ... }:
+
+with lib;
+
+{
+
+  mkCommonOptions = { name, defaults }: {
+
+    enable = mkEnableOption "Enable ${name}.";
+
+    package = mkOption {
+      description = "${name} package to use.";
+      type = types.package;
+      default = defaults.package;
+    };
+
+    executable = mkOption {
+      description = "Name of the executable within <option>package</option> to run.";
+      type = types.str;
+      default = defaults.executable or name;
+    };
+
+    name = mkOption {
+      description = "Name of the beat";
+      type = types.str;
+      default = defaults.name or name;
+    };
+
+    tags = mkOption {
+      description = "Tags to place on the shipped log messages";
+      type = types.listOf types.str;
+      default = [];
+    };
+
+    stateDir = mkOption {
+      description = "The state directory.";
+      type = types.str;
+      default = defaults.stateDir or "/var/lib/${name}";
+    };
+
+    loglevel = mkOption {
+      description = "Logging verbosity level.";
+      type = types.enum [ "debug" "info" "warning" "error" "fatal" ];
+      default = "warning";
+    };
+
+    extraConfig = mkOption {
+      description = "Additional config to be added to the beats config.json.";
+      type = types.attrs;
+      default = {};
+    };
+
+    elasticsearch = {
+      hosts = mkOption {
+        description = "Elasticsearch hosts";
+        type = with types; listOf str;
+        default = [ "localhost:9200" ];
+      };
+
+      username = mkOption {
+        description = "Username for elasticsearch basic auth.";
+        type = types.nullOr types.str;
+        default = null;
+      };
+
+      password = mkOption {
+        description = "Password for elasticsearch basic auth.";
+        type = types.nullOr types.str;
+        default = null;
+      };
+    };
+
+    kibana = {
+      host = mkOption {
+        description = "Host where kibana is reachable.";
+        type = types.str;
+        default = "localhost:5601";
+      };
+    };
+
+  };
+
+  mkNixosConfig = { cfg, mkBeatConfig }: let
+
+    mkCommonBeatConfig = cfg: {
+      inherit (cfg) tags;
+
+      path = {
+        data = "${cfg.stateDir}/data";
+        logs = "${cfg.stateDir}/logs";
+      };
+
+      logging.level = cfg.loglevel;
+
+      output = { inherit (cfg) elasticsearch; };
+
+      setup.template = {
+        fields = "${cfg.package}/share/fields.yml";
+        settings.index = {
+          number_of_shards = 1;
+          codec = "best_compression";
+        };
+      };
+
+      setup.kibana = {
+        inherit (cfg.kibana) host;
+      };
+
+      setup.dashboards = {
+        enabled = true;
+        directory ="${cfg.package}/share/kibana";
+      };
+    };
+
+    beatConfig = foldl' recursiveUpdate {} [
+      (mkCommonBeatConfig cfg)
+      (mkBeatConfig cfg)
+      cfg.extraConfig
```

Comment:
Since you're already doing something like that, I really suggest you do this instead: https://github.com/Infinisil/rfcs/blob/nixos-config/rfcs/0000-nixos-config.md#detailed-design, which will have a bunch of advantages

(ping @aanderse)

Reply:
Have opened an RFC since then for that: NixOS/rfcs#42